### ISSUE TITLE
Issue with get_uri when the encoding isn't 'utf-8' 

### ISSUE THREAD

**Issue by tumma72** [andreburgaud/robotspy#200]
The current implementation of gen_uri would fail with an entry in the robots.txt line which isn't correctly encoded in 'utf-8' additionally it is assuming that all robots.txt files are encoded in 'utf-8' which isn't a safe option. The following version solves the problem:

```python
def gen_uri(self, uri: str):
        """Instantiate a generator from a URI (either http:// or https:// or file:///"""
        try:
            with urllib.request.urlopen(uri) as f:
                self._time = time.time()
                encoding = f.headers.get_content_charset() or "utf-8"
                for line in f:
                    try:
                        yield line.decode(encoding)
                    except UnicodeDecodeError:
                        yield line.decode("utf-8", "replace")
        except urllib.error.HTTPError as err:
            if err.code in (401, 403):
                self.disallow_all = True
                self._errors.append((str(err.code), str(err)))
            elif 400 <= err.code < 500:
                self.allow_all = True
                self._warnings.append((str(err.code), str(err)))
            self._time = 0
        except urllib.error.URLError as err:
            self._errors.append(("", str(err)))
```

I have tried to submit a pull request but I have no rights 😉

**Comment by andreburgaud** [comment]
Hello @tumma72, Thank you for suggesting this change, and I am sorry for my late response. I tested it, and it works as expected. I'm not 100% sure we should accept a non-UTF-8 file as the RFC explicitly states that the file must be UTF-8 encoded https://www.rfc-editor.org/rfc/rfc9309#name-access-method. OTOH, the current code is failing, and so is `urllib.robotparser.RobotFileParser` in the Python standard library. To your point, we can't assume that a file is UTF-8 encoded, and it should fail gracefully and inform the user, if possible. I will address the issue and add the corresponding tests. Thank you again for spotting this problem.

**Comment by tumma72** [comment]
> Hello @tumma72, Thank you for suggesting this change, and I am sorry for my late response. I tested it, and it works as expected. I'm not 100% sure we should accept a non-UTF-8 file as the RFC explicitly states that the file must be UTF-8 encoded https://www.rfc-editor.org/rfc/rfc9309#name-access-method. OTOH, the current code is failing, and so is `urllib.robotparser.RobotFileParser` in the Python standard library. To your point, we can't assume that a file is UTF-8 encoded, and it should fail gracefully and inform the user, if possible. I will address the issue and add the corresponding tests. Thank you again for spotting this problem.

Cheers, as you say they should be UTF-8 but sometimes they aren't and some special characters slip through even if they shouldn't. My approach was attempting to recover as much as possible and eventually fail gracefully rather than interrupting the flow.

**Comment by andreburgaud** [comment]
Hi @tumma72, I applied your fix with a slight adjustment and published the package as version 0.8.0 along with a few minor updates. The change does not attempt to read a non-utf-8 file and behaves as if the `robots.txt` file did not exist, albeit it captures the error and makes it available to the user without crashing badly as it did before. Here is an example of the behavior, attempting to read a robots.txt file UTF-16 encoded:

```
$docker run --rm -it andreburgaud/robotspy https://robotspy.org/tests/robots_utf16.txt dummy-ua / 
user-agent 'RSPY' with path '/': ALLOWED
robots.txt must be UTF-8 encoded -> 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte for https://robotspy.org/tests/robots_utf16.txt
```
I'm going to close this issue. Thank you again for your help!

**Comment by andreburgaud** [comment]
Fixed in release 0.8.0 (https://pypi.org/project/robotspy/). Closing.
